linkerd/app/inbound/src/direct.rs

```diff
@@ -0,0 +1,135 @@
+use crate::{
+    http,
+    require_identity::RequireIdentityForDirect,
+    target::{RequestTarget, Target, TcpAccept, TcpEndpoint},
+};
+use linkerd_app_core::{
+    config::ProxyConfig,
+    detect, drain, io, metrics,
+    opencensus::proto::trace::v1 as oc,
+    proxy::identity::LocalCrtKey,
+    svc, tls,
+    transport::{listen, metrics::SensorIo},
+    transport_header::{DetectHeader, TransportHeader},
+    Error, DST_OVERRIDE_HEADER,
+};
+use std::fmt::Debug;
+use tokio::sync::mpsc;
+
+#[derive(Default)]
+struct NonOpaqueRefused(());
+
+type FwdIo<I> = io::PrefixedIo<SensorIo<tls::server::Io<I>>>;
+
+/// Builds a stack that handles connections that target the proxy's inbound port
+/// (i.e. without an SO_ORIGINAL_DST setting). This port behaves differently from
+/// the main proxy stack:
+///
+/// 1. Protocol detection is always performed;
+/// 2. TLS is required;
+/// 3. A transport header is expected. It's not strictly required, as
+///    gateways may need to accept HTTP requests from older proxy versions
+pub fn stack<I, F, FSvc, L, LSvc>(
+    config: &ProxyConfig,
+    local_identity: Option<LocalCrtKey>,
+    tcp_forward: F,
+    http_loopback: L,
+    metrics: &metrics::Proxy,
+    span_sink: Option<mpsc::Sender<oc::Span>>,
+    drain: drain::Watch,
+) -> impl svc::NewService<
+    listen::Addrs,
+    Service = impl svc::Service<I, Response = (), Error = Error, Future = impl Send>,
+> + Clone
+where
+    I: io::AsyncRead + io::AsyncWrite + io::Peek + io::PeerAddr,
+    I: Debug + Send + Sync + Unpin + 'static,
+    F: svc::NewService<TcpEndpoint, Service = FSvc> + Clone + Send + Sync + Unpin + 'static,
+    FSvc: svc::Service<FwdIo<I>, Response = ()> + Clone + Send + Sync + Unpin + 'static,
+    FSvc::Error: Into<Error>,
+    FSvc::Future: Send + Unpin,
+    L: svc::NewService<Target, Service = LSvc> + Clone + Send + Sync + Unpin + 'static,
+    LSvc: svc::Service<http::Request<http::BoxBody>, Response = http::Response<http::BoxBody>>
+        + Send
+        + 'static,
+    LSvc::Error: Into<Error>,
+    LSvc::Future: Send,
+{
+    let detect_timeout = config.detect_protocol_timeout;
+    let dispatch_timeout = config.dispatch_timeout;
+
+    // Direct traffic may target an HTTP gateway.
+    let http = svc::stack(http_loopback)
+        .push_on_response(
+            svc::layers()
+                .push(svc::FailFast::layer("Gateway", dispatch_timeout))
+                .push_spawn_buffer(config.buffer_capacity),
+        )
+        .check_new_service::<Target, http::Request<http::BoxBody>>()
+        // Removes the override header after it has been used to
+        // determine a request target.
+        .push_on_response(http::strip_header::request::layer(DST_OVERRIDE_HEADER))
+        // Routes each request to a target, obtains a service for that
+        // target, and dispatches the request.
+        .instrument_from_target()
+        .push(svc::NewRouter::layer(RequestTarget::from))
```

Comment:
TIOLI: does it make sense to bundle the router layer, strip header, and instrument layers together somewhere? the direct and http router stacks both have the same configuration.

Reply:
It's going to change on the direct stack, so no.

In general: [rule of three](https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming)).